### PR TITLE
Strip windows line endings

### LINKED ISSUE #107: Template don't compile with CRLF line endings ie. Windows
When you have template like this

```
doctype html
html
  head
    title Hello Rumbl!
  body
   Hello
```

With standard CRLF windows endings you get:

```
== Compilation error on file web/views/layout_view.ex ==
** (FunctionClauseError) no function clause matching in Slime.Doctype.for/1
    lib/slime/doctype.ex:6: Slime.Doctype.for("html\r")
    lib/slime/parser.ex:200: Slime.Parser.parse_line/2
    lib/slime/parser.ex:75: Slime.Parser.parse_line/1
    lib/slime/parser.ex:47: Slime.Parser.parse_lines/2
    lib/slime/renderer.ex:20: Slime.Renderer.precompile/1
    lib/phoenix_slime/engine.ex:9: PhoenixSlime.Engine.compile/2
    (phoenix) lib/phoenix/template.ex:378: Phoenix.Template.compile/2
    (phoenix) lib/phoenix/template.ex:186: Phoenix.Template."-MACRO-__before_compile__/2-fun-0-"/3
```

And I don't think this being phoenix app is relevant.

When you change line endings to standard Linux ones: LF, the compilation on the same machine does not give you any error and it simply works.

### PR DESCRIPTION
**Pull request by tmbb**
All credit goes to sztosz, who fixed this issue originally.
Added test.

Fixes https://github.com/slime-lang/slime/issues/107

### CODE REVIEW

**Comment by Rakoth** [comment]
Hi, @tmbb could you please provide more details about the issue, parser now should handle windows line ending, we have corresponding test case: 

https://github.com/slime-lang/slime/blob/master/test/renderer_test.exs#L91-L93
```
  test "CRLF line endings are converted to LF" do
    assert render("h1\r\n\th2\r\n\t\th3 Hi\r\n") == "<h1><h2><h3>Hi</h3></h2></h1>"
  end
```

**Comment by tmbb** [comment]
That's strange... I admit I didn't read the whole source, but the master failed in my machine before my fix and started working with my fix. I'll try to figure it out.

**Comment by tmbb** [comment]
So far:

* That test passes in my machine (i.e. `mix tests` passes)
* The project I'm trying this on, which uses phoenix slime, fails with this error:

```
== Compilation error on file lib/talon_blog_demo/talon/views/admin/admin-lte/layout_view.ex ==
** (Slime.TemplateSyntaxError) got FunctionClauseError with message "no function clause matching in String.duplicate/2" while retrieving Exception.message/1 for %Slime.TemplateSyntaxError{column: 0, line: "  head\r", line_number: 3, message: "{:no_match, <<14>>}", source: "INPUT"}
    (slime) lib/slime/parser.ex:27: Slime.Parser.handle_syntax_error/3
    (slime) lib/slime/renderer.ex:17: Slime.Renderer.precompile/1
    (phoenix_slime) lib/phoenix_slime/engine.ex:11: PhoenixSlime.Engine.compile/2
    (phoenix) lib/phoenix/template.ex:378: Phoenix.Template.compile/2
    (phoenix) lib/phoenix/template.ex:186: Phoenix.Template."-MACRO-__before_compile__/2-fun-0-"/3
    (elixir) lib/enum.ex:1755: Enum."-reduce/3-lists^foldl/2-0-"/3
    (phoenix) expanding macro: Phoenix.Template.__before_compile__/1
    lib/talon_blog_demo/talon/views/admin/admin-lte/layout_view.ex:1: TalonBlogDemo.Admin.AdminLte.Web.LayoutView (module)
    (elixir) lib/kernel/parallel_compiler.ex:117: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/1
```

Not that the final error comes from trying to call String.duplicate(" ", -1), as the error reported at column 0. This is strange (the code doesn't seem to think that an exception at column 0 is possible), but that's not the most important part. The most important part is the syntax error.

I'm not yet deep enough into Slime's source, but the error message ("{:no_match, <<14>>}") seems to indicate an unexpected "\r" character.

The repo of the project is this: https://github.com/talonframework/talon_blog_demo

**Steps to repeat (on windows):**

* git clone https://github.com/talonframework/talon_blog_demo
* cd talon_blog_demo
* mix deps.get
* mix

---

The problem goes away with my fix.
My fix works at the level of the preprocessor.
I don't know what's the relevance of that fact...

The problem also goes away with if {:slime, "~> 0.16.0"} is used instead of 1.0.0.

Thanks for taking the time to work on this issue.

**Comment by Rakoth** [comment]
I see, the problem is with preprocessor, which inserts virtual symbols of indent `<<14>>` and dedent `<<15>>`, but since preprocessor splits lines by `\n` it inserts one of them between `\r` and `\n`.
Could you narrow it down to exact slime template, which causes this problem, so we can add a test case?

**Comment by tmbb** [comment]
I haven't done any formal analysis to try to find out which templates break Slime on windows, but I've managed to shrink my test case into a possibly minimal example.

It would be interesting to add property testing to Slime. It would have caught this bug in testing (note, that you already had a test case for CRLF problems, but only a slightly more complex template found the error; such more complex templates could be generated as a matter of by property testing frameworks).

If I find the time to learn something like [propcheck](https://github.com/alfert/propcheck) (the custom element generators seem great), I can submit a pull request.

File: test/renderer_test.exs [hunk 1]
```diff
@@ -91,4 +91,18 @@ defmodule RendererTest do
...
+
+  test "CRLF line endings corner case" do
+    example_unix = """
+    html
+      head
+        meta
+
+      body
+    """
+    example_windows = example_unix |> String.replace("\n", "\r\n")
```
Comment (Rakoth) [inline test/renderer_test.exs]: I am not sure, this test will do the same on windows machine, I think, it is better to state all symbols in a string as in previous test

File: test/renderer_test.exs [hunk 2]
```diff
@@ -91,4 +91,18 @@ defmodule RendererTest do
...
+
+    assert Slime.Renderer.precompile(example_unix) == Slime.Renderer.precompile(example_windows)
+    assert render(example_unix) == render(example_unix)
```
Comment (Rakoth) [inline test/renderer_test.exs]: better to compare to expected html string (also typo)

**Review by Rakoth — CHANGES_REQUESTED** [review]

**Comment by Rakoth** [comment]
Ok, I see the problem now, it is with empty line in template.
Sure @tmbb you can send a PR with propcheck 👍

File: test/renderer_test.exs (hunk 1 above)
Comment (tmbb) [reply test/renderer_test.exs]: It doesn't do the same: it fails on v1.0.0 with an error. Independently of the OS, elixir normalizes the line endings in multiline strings. Line endings will always be "\n", both in windows and linux.

By replacing the linux line endings with windows line endings you mimic the effect of running slime on windows files.

I can compare with the expected output, of course, if you want.

File: test/renderer_test.exs (hunk 2 above)
Comment (tmbb) [reply test/renderer_test.exs]: Which typo? I can't find it

Comment (tmbb) [reply test/renderer_test.exs]: Sorry, found it

File: test/renderer_test.exs (hunk 1 above)
Comment (Rakoth) [reply test/renderer_test.exs]: Nice to know, thanks 👍

**Review by Rakoth — APPROVED** [review]
